### PR TITLE
simplify implementation of memory efficient + GPT-OSS loading

### PR DESCRIPTION
**Pull request by RobotSail**
This PR:
- decouples GPT-OSS from OSFT memory efficient loading
- ensures that only the intended params are loaded on the GPU during memory efficient loading
- ensures all param types are converted to the intended `train_dtype` regardless of the training mode
- adds epoch to the logger

### CODE REVIEW

File: src/mini_trainer/osft_utils.py [hunk 1]
```diff
@@ -394,6 +394,51 @@ def reconstruct_weight_matrix(
     return reconstructed
 
 
+def broadcast_svd_results(model, assignments, world_size):
```
Comment (NikhilNayak-debug) [inline src/mini_trainer/osft_utils.py]: We already have this method defined in this file later on.

File: src/mini_trainer/osft_utils.py [hunk 2]
```diff
@@ -763,60 +815,50 @@ def _load_model_memory_efficient(
         osft_memory_efficient_init=osft_memory_efficient_init
     )
     
-    # Load state dict into OSFT model
-    log_rank_0("📋 Loading state dict into OSFT model...")
-    model.load_state_dict(state_dict, strict=False)
+    if osft_memory_efficient_init:
```
Comment (NikhilNayak-debug) [inline src/mini_trainer/osft_utils.py:772]: So if I understand correctly the SVD params and buffers are created on CPU and then moved to GPU later on along with the remaining frozen params in `wrap_fsdp2` method. Earlier we were creating the SVD params and buffers on each GPU and only the remaining frozen params were moved to GPU in `wrap_fsdp2` method. So now we save from having all SVD params on each gpu because they are sharded later in `wrap_fsdp2` while moving to GPU. Nice!

File: src/mini_trainer/setup_model_for_training.py [hunk 3]
```diff
@@ -165,6 +165,93 @@ def get_model_save_dtype(save_dtype: str | torch.dtype | None, model_name_or_pat
     return save_dtype
 
 
+def setup_osft_model(
```
Comment (NikhilNayak-debug) [inline src/mini_trainer/setup_model_for_training.py:168]: Do we want this method in this file or have it in the `osft_utils` file instead and have it called from the setup_model method?

**Review by NikhilNayak-debug — COMMENTED** [review]
Changes look good to me, now we are more memory efficient!

File: src/mini_trainer/osft_utils.py (hunk 1 above)
Comment (RobotSail) [reply src/mini_trainer/osft_utils.py]: Yeah this was added by mistake

File: src/mini_trainer/setup_model_for_training.py (hunk 3 above)
Comment (RobotSail) [reply src/mini_trainer/setup_model_for_training.py:168]: @NikhilNayak-debug I left a comment on the other thread about this -- we don't want to keep this method in osft_utils.py if importing it from there will cause circular dependencies.

In this case, this method only calls other functions from that file, but is not used by anything there. So it's fine to keep here.

Comment (NikhilNayak-debug) [reply src/mini_trainer/setup_model_for_training.py:168]: Sounds good to me!

File: src/mini_trainer/osft_utils.py (hunk 2 above)
Comment (RobotSail) [reply src/mini_trainer/osft_utils.py:772]: Yeah, another problem was also that we had coupled the optimizations to GPT-OSS early on. So additional effort was required to decouple them, so the correct logic is only the SVD params go onto the GPU device gradually instead of the whole thing.

During testing, memory usage on each node appears to grow towards around ~6.5-7GB/gpu, across 8 cards, or around 56GB overall. This is similar to what we saw previously, where usage would grow to ~70GB/card during init for an 8B param model.

Comment (RobotSail) [reply src/mini_trainer/osft_utils.py:772]: @NikhilNayak-debug I'm not sure why, but now the memory efficient loading is much faster after these changes, about the same as the memory inefficient method. Probably because we are still keeping stuff on GPU, just not everything all at once.
